### ISSUE TITLE
`write!` is somehow masking the `#[must_use]` on its result

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#55516]
Hat tip to @shepmaster for finding this

[play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=267a47eba842626f4cce618ad89caa47)

```rust
fn main() {
    use std::fmt::Write;
    let mut example = String::new();
    write!(&mut example, "{}", 42);
}
```

is not triggering the `#[must_use]` lint on `Result`.

**Comment by shepmaster** [comment]
This code compiles with warnings in Rust 1.28.0, but no warnings in Rust 1.29.0

```rust
fn main() {
    use std::fmt::Write;
    let mut example = String::new();
    write!(&mut example, "{}", 42);
}
```

```
$ cargo +1.28.0 build
   Compiling xz v0.1.0 (file:///private/tmp/xz)
warning: unused `std::result::Result` which must be used
 --> src/main.rs:4:5
  |
4 |     write!(&mut example, "{}", 42);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_must_use)] on by default
  = note: this `Result` may be an `Err` variant, which should be handled
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

    Finished dev [unoptimized + debuginfo] target(s) in 1.02s

$ cargo clean
$ cargo +1.29.0 build
   Compiling xz v0.1.0 (file:///private/tmp/xz)
    Finished dev [unoptimized + debuginfo] target(s) in 1.09s
```

Potentially related to #54288.

**Comment by shepmaster** [comment]
Expanding the macro and compiling does have the warning:

```rust
fn main() {
    use std::fmt::Write;
    let mut example = String::new();
    (&mut example).write_fmt(format_args!("{}", 42));
}
```

```
$ cargo clean
$ cargo +1.29.0 build
   Compiling xz v0.1.0 (file:///private/tmp/xz)
warning: unused `std::result::Result` which must be used
 --> src/main.rs:4:5
  |
4 |     (&mut example).write_fmt(format_args!("{}", 42));
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_must_use)] on by default
  = note: this `Result` may be an `Err` variant, which should be handled

    Finished dev [unoptimized + debuginfo] target(s) in 0.56s
```

**Comment by ehuss** [comment]
Seems related to #52467.

**Comment by shepmaster** [comment]
> related to #52467.

/cc @alexcrichton

**Comment by alexcrichton** [comment]
Ah oops, definitely a mistake! I believe a fix for this would be to set `report_in_external_macro` ([like in this macro](https://github.com/rust-lang/rust/blob/4bd4e4130ed531a644263db26bf8461704215c77/src/librustc/lint/builtin.rs#L77-L82)) for the [definition of the unused must use lint](https://github.com/rust-lang/rust/blob/e8aef7cae14bc7a56859408c90253e9bcc07fcff/src/librustc_lint/unused.rs#L29-L33) as well

**Comment by varkor** [comment]
Will https://github.com/rust-lang/rust/issues/55240 also be fixed by this?
